framework/src/modules/chain/submodules/blocks/process.js

```diff
@@ -427,11 +435,242 @@ class Process {
+			if (
+				forgingSlotLastBlock < forgingSlotNewBlock &&
+				!this._receivedInSlot(lastBlock, lastBlockReceivedAt) &&
+				this._receivedInSlot(block, newBlockReceivedAt)
+			) {
+				// Case 4: Tie break
+				return this._handleDoubleForgingTieBreak(block, lastBlock);
+			}
+		} else if (
+			lastBlock.heightPrevoted < block.heightPrevoted ||
+			(lastBlock.height < block.height &&
+				lastBlock.heightPrevoted === block.heightPrevoted)
+		) {
+			// Case 5: received block has priority. Move to a different chain.
+
+			return this._handleMovingToDifferentChain();
+			// TODO: Move to a different chain
+		}
+
+		// Discard newly received block
+		return library.logger.warn(
+			`Discarded block that does not match with current chain: ${
+				block.id
+			} height: ${block.height} round: ${slots.calcRound(
+				block.height
+			)} slot: ${slots.getSlotNumber(block.timestamp)} generator: ${
+				block.generatorPublicKey
+			}`
+		);
+	}
+
+	/**
+	 * Block IDs are the same ~ Blocks are equal
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleSameBlockReceived(block) {
+		library.logger.debug('Block already processed', block.id);
+	}
+
+	/**
+	 * Block received is correct
+	 * @param block
+	 * @returns {Promise}
+	 * @private
+	 */
+	_handleGoodBlock(block) {
+		return this._processBlock(block);
+	}
+
+	/**
+	 * Double forging. Last block stands
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDoubleForging(block) {
+		library.logger.warn(
+			'Delegate forging on multiple nodes',
+			block.generatorPublicKey
+		);
+		library.logger.info('Last block stands');
+		// TODO: Implement Proof of Misbehavior
+	}
+
+	/**
+	 * Tie break: two competing blocks by different delegates at the same height.
+	 * @param lastBlock
+	 * @param newBlock
+	 * @returns {Promise}
+	 * @private
+	 */
+	async _handleDoubleForgingTieBreak(newBlock, lastBlock) {
+		try {
+			// Normalize and check schema
+			const normalizedNewBlock = library.logic.block.objectNormalize(
+				_.cloneDeep(newBlock)
+			);
+
+			// Validate block has been generated by correct delegate
+			await this._validateGenerator(normalizedNewBlock, lastBlock);
+
+			// Verify block data, signatures, etc
+			// TODO: Maybe `verifyReceipt` could throw errors instead of returning checks.
+			const check = modules.blocks.verify.verifyReceipt(normalizedNewBlock);
+
+			if (!check.verified) {
+				library.logger.error(
+					`Block ${normalizedNewBlock.id} verification failed`,
+					check.errors.join(', ')
+				);
+				// Return first error from checks
+				throw check.errors[0];
+			}
+		} catch (error) {
+			// Should there be any errors while validating and verifying,
+			// last block is kept as the tip of the chain
+			library.logger.error(
+				'Case 4 fork recovery failed',
+				convertErrorsToString(error)
+			);
+			throw error;
+		}
```

Comment:
Don't we already have this validation logic? Can't we use `modules.blocks.verify.procssBlock` here?

Reply:
`this._processBlock()` call here is actually executing the old function:

```
__private.receiveBlock = function(block, cb) {
	library.logger.info(
		`Received new block id: ${block.id} height: ${
			block.height
		} round: ${slots.calcRound(block.height)} slot: ${slots.getSlotNumber(
			block.timestamp
		)} reward: ${block.reward}`
	);

	// Update last receipt
	modules.blocks.lastReceipt.update(); // TODO: Probably not needed anymore.
	// Start block processing - broadcast: true, saveBlock: true
	modules.blocks.verify.processBlock(block, true, true, cb);
};
```

Which is intended to be refactored and removed from _private after compatibility window is over. It is not only calling modules.blocks.verify.processBlock but also logging and updating lastReceipt as you can see above.

Reply:
@shuse2 is working on this right?

Reply:
I'm waiting for that PR to make my changes as it will cause a lot of conflicts.